polygeist/lib/polygeist/Passes/ConvertPolygeistToLLVM.cpp

```diff
@@ -90,84 +84,131 @@ struct SubIndexOpLowering : public ConvertOpToLLVMPattern<SubIndexOp> {
-               "type");
-      } else
-        assert(sycl::isSYCLType(sourceMemRefType.getElementType()) &&
-               "Expecting element type of the memref to be a SYCL type");
-
-      // The first index (zero) takes a pointer to the structure.
-      Value zero = rewriter.create<LLVM::ConstantOp>(
-          loc, idxs[0].getType(),
-          rewriter.getIntegerAttr(idxs[0].getType(), 0));
-      Value idxs[] = {zero, transformed.index()};
-
-      // According to MLIRScanner::InitializeValueByInitListExpr() in
-      // clang-mlir.cc, when a memref element type is a struct type, the return
-      // type of a polygeist.subindex should be a memref of the element type of
-      // the struct.
-      auto elemPtrTy = LLVM::LLVMPointerType::get(
-          getTypeConverter()->convertType(viewMemRefType.getElementType()));
-      auto gep = rewriter.create<LLVM::GEPOp>(loc, elemPtrTy, prev, idxs);
-      MemRefDescriptor nexRef = createMemRefDescriptor(
-          loc, viewMemRefType, gep, gep, sizes, strides, rewriter);
-
-      rewriter.replaceOp(subViewOp, {nexRef});
+    Type sourceElemType = sourceMemRefType.getElementType();
+    Type convSourceElemType = getTypeConverter()->convertType(sourceElemType);
+    Type viewElemType = viewMemRefType.getElementType();
+    Type convViewElemType = getTypeConverter()->convertType(viewElemType);
+
+    // When the converted source operand element type is a struct, the original
+    // memref element type represents a SYCL type. Handle the non-SYCL case
+    // first.
+    if (!convSourceElemType.isa<LLVM::LLVMStructType>()) {
+      assert(
+          convViewElemType ==
+              prev.getType().cast<LLVM::LLVMPointerType>().getElementType() &&
+          "Expecting the element types to match");
+      auto memRefDesc = createMemRefDescriptor(
+          loc, viewMemRefType, targetMemRef.allocatedPtr(rewriter, loc),
+          rewriter.create<LLVM::GEPOp>(loc, prev.getType(), prev, idxs), sizes,
+          strides, rewriter);
+
+      rewriter.replaceOp(subViewOp, {memRefDesc});
       return success();
     }
 
-    assert(getTypeConverter()->convertType(viewMemRefType.getElementType()) ==
-               prev.getType().cast<LLVM::LLVMPointerType>().getElementType() &&
-           "Expecting the element types to match");
-    MemRefDescriptor nexRef = createMemRefDescriptor(
-        loc, viewMemRefType, targetMemRef.allocatedPtr(rewriter, loc),
-        rewriter.create<LLVM::GEPOp>(loc, prev.getType(), prev, idxs), sizes,
-        strides, rewriter);
-
-    rewriter.replaceOp(subViewOp, {nexRef});
+    // SYCL case
+    assert(sourceMemRefType.getRank() == viewMemRefType.getRank() &&
+           "Expecting the input and output MemRef ranks to be the same");
+    // Note:in MLIRASTConsumer::getMLIRType(), a memref of struct type is
+    // generated only for a struct that has at least one member of SYCL type,
+    // otherwise a llvm pointer type is generated instead of a memref).
```

Comment:
```suggestion
    // Note: in MLIRASTConsumer::getMLIRType(), a memref of struct type is
    // generated only for a struct that has at least one member of SYCL type,
    // otherwise a llvm pointer type is generated instead of a memref.
```

Reply:
done

Reply:
still need to remove the `)` at the end of memref.

Reply:
darn missed it

Reply:
done